src/chainrules.jl

```diff
@@ -38,14 +44,24 @@ end
 function _Ω_∂_xlog1py(x::Real, y::Real)
     log1py = log1p(y)
     z = x * log1py
-    Ω = iszero(x) && !isnan(y) ? zero(z) : z
+    w = x / (1 + y)
+    if iszero(x) && !isnan(y)
+        Ω = zero(z)
+        ∂y = zero(w)
+    else
+        Ω = z
+        ∂y = w
+    end
     ∂x = log1py
-    ∂y = x / (1 + y)
     return Ω, ∂x, ∂y
 end
 function ChainRulesCore.frule((_, Δx, Δy), ::typeof(xlog1py), x::Real, y::Real)
     Ω, ∂x, ∂y = _Ω_∂_xlog1py(x, y)
-    ΔΩ = muladd(∂x, Δx, ∂y * Δy)
+    if iszero(Δx)
+        ΔΩ = ∂y * Δy
+    else
+        ΔΩ = muladd(∂x, Δx, ∂y * Δy)
+    end
```

Comment:
Can you explain this change? Maybe you did but I'm still confused. Why do we want that `Δx` acts always like a strong zero? Shouldn't that only be the case when we pass `ZeroTangent` etc. or `false`?

Reply:
Well, otherwise the `-Inf` will result in `ΔΩ` being a `NaN`, which is different from what the `rrule` produces. Actually if the policy is to avoid NaNs, using `muladd` seems to make this harder in general. But I don't have a very good understanding of these matters as they relate to other considerations in the ecosystem. Maybe we can ask someone else for a second review.

Reply:
That's a general issue in ChainRules currently (there exist some issues as well IIRC). If you want to avoid `NaN`s usually you have to pass `ZeroTangent`/`NoTangent`/some other `AbstractTangent`/`false`, i.e., anything that is a strong zero. For instance:
```julia
julia> using ChainRules, ChainRulesCore

julia> frule((NoTangent(), 0.0), log, 0.0)
(-Inf, NaN)

julia> frule((NoTangent(), ZeroTangent()), log, 0.0)
(-Inf, ZeroTangent())

julia> frule((NoTangent(), NoTangent()), log, 0.0)
(-Inf, NoTangent())

julia> frule((NoTangent(), false), log, 0.0)
(-Inf, 0.0)

julia> frule((NoTangent(), 0.0), log1p, -1.0)
(-Inf, NaN)

julia> frule((NoTangent(), ZeroTangent()), log1p, -1.0)
(-Inf, ZeroTangent())

julia> frule((NoTangent(), NoTangent()), log1p, -1.0)
(-Inf, NoTangent())

julia> frule((NoTangent(), false), log1p, -1.0)
(-Inf, 0.0)
```

Reply:
I see, thanks for clearing that up. So I think it is better to revert this change and use `ZeroTangent()` in the test.